File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        match = {'(':')', '{':'}', '[':']'} #dictionary
        
        for c in s:
            if c in match:
                stack.append(c)
            else:
                if len(stack)==0: return False
                if match[stack.pop()]!=c: return False
            
        return len(stack)==0

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        char c;
        for(int i=0; i<s.length(); i++){
            if(s[i]=='(' || s[i]=='{'|| s[i]=='[')
                st.push(s[i]);
            else if(s[i]==')' || s[i]=='}'|| s[i]==']') {
                if(st.empty()) return false;
                c = st.top(); st.pop();
                if(c=='(' && s[i]!=')') return false;
                if(c=='{' && s[i]!='}') return false;
                if(c=='[' && s[i]!=']') return false;
            }
        }
        if(!st.empty()) return false; //exception
        return true;
    }
};